src/test/resources/refaster/PreConditionsVerifierRecipes.java

```diff
@@ -0,0 +1,403 @@
+            return Preconditions.check(
+                    new UsesMethod<>("java.io.PrintStream println(..)", true),
+                    javaVisitor
+            );
+        }
+    }
+
+    /**
+     * OpenRewrite recipe created for Refaster template {@code PreConditionsVerifier.NoUsesTypeWhenBeforeTemplateContainsPrimitiveAndAnotherType}.
+     */
+    @SuppressWarnings("all")
+    @NullMarked
+    @Generated("org.openrewrite.java.template.processor.RefasterTemplateProcessor")
+    public static class NoUsesTypeWhenBeforeTemplateContainsPrimitiveAndAnotherTypeRecipe extends Recipe {
+
+        /**
+         * Instantiates a new instance.
+         */
+        public NoUsesTypeWhenBeforeTemplateContainsPrimitiveAndAnotherTypeRecipe() {}
+
+        @Override
+        public String getDisplayName() {
+            return "Refaster template `PreConditionsVerifier.NoUsesTypeWhenBeforeTemplateContainsPrimitiveAndAnotherType`";
+        }
+
+        @Override
+        public String getDescription() {
+            return "Recipe created for the following Refaster template:\n```java\npublic static class NoUsesTypeWhenBeforeTemplateContainsPrimitiveAndAnotherType {\n    \n    @BeforeTemplate()\n    void before(int actual) {\n        System.out.println(actual);\n    }\n    \n    @BeforeTemplate()\n    void before(Map<?, ?> actual) {\n        System.out.println(actual);\n    }\n    \n    @AfterTemplate()\n    void after(Object actual) {\n        System.out.println(\"Changed: \" + actual);\n    }\n}\n```\n.";
+        }
+
+        @Override
+        public TreeVisitor<?, ExecutionContext> getVisitor() {
+            JavaVisitor<ExecutionContext> javaVisitor = new AbstractRefasterJavaVisitor() {
+                final JavaTemplate before = JavaTemplate
+                        .builder("System.out.println(#{actual:any(int)});")
+                        .build();
+                final JavaTemplate before0 = JavaTemplate
+                        .builder("System.out.println(#{actual:any(java.util.Map<?,?>)});")
+                        .build();
+                final JavaTemplate after = JavaTemplate
+                        .builder("System.out.println(\"Changed: \" + #{actual:any(java.lang.Object)});")
+                        .build();
+
+                @Override
+                public J visitMethodInvocation(J.MethodInvocation elem, ExecutionContext ctx) {
+                    JavaTemplate.Matcher matcher;
+                    if ((matcher = before.matcher(getCursor())).find()) {
+                        return embed(
+                                after.apply(getCursor(), elem.getCoordinates().replace(), matcher.parameter(0)),
+                                getCursor(),
+                                ctx,
+                                SHORTEN_NAMES
+                        );
+                    }
+                    if ((matcher = before0.matcher(getCursor())).find()) {
+                        maybeRemoveImport("java.util.Map");
```

Comment:
Should this stay? I think so, right? _If I am correct, the `maybeRemoveImport` is smart enough to figure out whether the import should be removed or not._

Reply:
Indeed; imports will only be removed if there are no more LST elements using those types. In addition to that the `before0.matcher` already indicated we're only handling cases where there is a Map being matched.